### ISSUE TITLE
Function keyword for identation

### ISSUE THREAD

**Issue by lzap** [vim/vim#16049]
### Steps to reproduce

1. Create a file `vim -N test.sh`
2. Turn on `set smartindent` and `set ft=sh` or `set ft=bash`
3. Type `function a-function {`
4. Identation is not correct, function is not recognized.

### Expected behaviour

Function is recognized, although I am not entirely sure.

The [bash docs mention](https://www.gnu.org/software/bash/manual/bash.html) it can be any "shell word", but does not explain this term further:

When the shell is in POSIX mode (see Bash POSIX Mode), fname must be a valid shell name and may not be the same as one of the special builtins (see Special Builtins). In default mode, a function name can be any unquoted shell word that does not contain ‘$’. Any redirections (see Redirections) associated with the shell function are performed when the function is executed. A function definition may be deleted using the -f option to the unset builtin (see Bourne Shell Builtins).

In Bash, such function works. I wonder if would be possible to override the function definition for `bash.vim`:

```
function! s:is_function_definition(line)
  return a:line =~ '^\s*\<\k\+\>\s*()\s*{' ||
       \ a:line =~ '^\s*{' ||
       \ a:line =~ '^\s*function\s*\k\+\s*\%(()\)\?\s*{'
endfunction
```

A you can see it currently only recognize "kewords" for function names.

### Version of Vim

Vi IMproved 9.1.825

### Environment

Fedora 41 stock `vim-enhanced`

**Comment by chrisbra** [comment]
> Turn on set smartindent and set ft=sh or set ft=bash

That looks wrong. The default shell indentexpr seems to handle your case just fine.

**Comment by lzap** [comment]
Yeah that is exactly what I think, tho, I think this needs to be overriden in `bash.vim` which currently includes `sh.vim`:

```
sh-5.2$ function test-it() {
> echo "test"
> }
sh: „test-it“: not valid identifier
sh-5.2$ 
exit

lzap@nuc:~$ function test-it() {
> echo "test"
> }
```

I will try to send a patch, thank you.

**Comment by chrisbra** [comment]
In that case, we probably need a patch like this:
```diff
diff --git a/runtime/indent/sh.vim b/runtime/indent/sh.vim
index aa47c6d1b..ac4bbef1f 100644
--- a/runtime/indent/sh.vim
+++ b/runtime/indent/sh.vim
@@ -195,7 +195,9 @@ endfunction
 function! s:is_function_definition(line)
   return a:line =~ '^\s*\<\k\+\>\s*()\s*{' ||
        \ a:line =~ '^\s*{' ||
-       \ a:line =~ '^\s*function\s*\k\+\s*\%(()\)\?\s*{'
+       \ a:line =~ '^\s*function\s*\k\+\s*\%(()\)\?\s*{' ||
+       \ (get(b:, 'is_bash', 0) &&
+       \  a:line =~ '^\s*function\s*\S\+\s*\%(()\)\?\s*{' )
 endfunction

 function! s:is_array(line)
```

**Comment by lzap** [comment]
https://github.com/vim/vim/pull/16052

**Comment by chrisbra** [comment]
fixed by 0acd3abfadf54a97a3cad999791603aa6a8b2248
